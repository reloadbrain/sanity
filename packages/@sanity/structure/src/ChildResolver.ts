import {StructureNode, EditorNode} from './StructureNodes'

export interface ChildResolverOptions {
  index: number
}

export interface ChildResolver {
  (itemId: string, parent: StructureNode, options: ChildResolverOptions):
    | StructureNode
    | EditorNode
    | Promise<StructureNode | EditorNode>
    | undefined
}
