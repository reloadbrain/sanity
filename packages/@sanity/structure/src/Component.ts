import {SerializeOptions, StructureNode, Serializable} from './StructureNodes'
import {SerializeError, HELP_URL} from './SerializeError'
import {MenuItem, MenuItemBuilder, maybeSerializeMenuItem} from './MenuItem'
import {MenuItemGroup, MenuItemGroupBuilder, maybeSerializeMenuItemGroup} from './MenuItemGroup'

export interface Component extends StructureNode {
  component: Function
  menuItems: MenuItem[]
  menuItemGroups: MenuItemGroup[]
}

export interface ComponentInput extends StructureNode {
  component: Function
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export interface BuildableList extends Partial<StructureNode> {
  component?: Function
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export class ComponentBuilder implements Serializable {
  protected spec: BuildableList

  constructor(spec?: ComponentInput) {
    this.spec = spec ? spec : {}
  }

  id(id: string) {
    this.spec.id = id
    return this
  }

  title(title: string) {
    this.spec.title = title
    return this
  }

  component(component: Function) {
    this.spec.component = component
    return this
  }

  menuItems(items: (MenuItem | MenuItemBuilder)[]) {
    this.spec.menuItems = items
    return this
  }

  menuItemGroups(groups: (MenuItemGroup | MenuItemGroupBuilder)[]) {
    this.spec.menuItemGroups = groups
    return this
  }

  serialize(options: SerializeOptions = {path: []}): Component {
    const {id, title, component} = this.spec
    if (typeof id !== 'string' || !id) {
      throw new SerializeError(
        '`id` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!component) {
      throw new SerializeError(
        '`component` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    return {
      id,
      title,
      type: 'component',
      component,
      menuItems: (this.spec.menuItems || []).map((item, i) =>
        maybeSerializeMenuItem(item, i, options.path)
      ),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) =>
        maybeSerializeMenuItemGroup(item, i, options.path)
      )
    }
  }
}
