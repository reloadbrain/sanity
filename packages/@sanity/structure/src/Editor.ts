import {EditorNode} from './StructureNodes'

export type PartialEditorNode = {
  id?: string
  title?: string
  options?: {
    id?: string
    type?: string
  }
}

export class EditorBuilder {
  protected spec: PartialEditorNode

  constructor(spec?: PartialEditorNode) {
    this.spec = spec || {}
  }

  id(id: string): EditorBuilder {
    this.spec.id = id
    return this
  }

  title(title: string): EditorBuilder {
    this.spec.title = title
    return this
  }

  documentId(documentId: string): EditorBuilder {
    this.spec.options = {
      ...(this.spec.options || {}),
      id: documentId
    }

    return this
  }

  type(documentType: string): EditorBuilder {
    this.spec.options = {
      ...(this.spec.options || {}),
      type: documentType
    }

    return this
  }

  serialize(): EditorNode {
    const {id, options} = this.spec
    if (typeof id !== 'string' || !id) {
      throw new Error('`id` is required for editor nodes')
    }

    if (!options || !options.id) {
      throw new Error('document id (`id`) is required for editor nodes')
    }

    if (!options || !options.type) {
      throw new Error('document type (`type`) is required for editor nodes')
    }

    return {
      ...this.spec,
      id,
      type: 'document',
      options: {id: options.id, type: options.type}
    }
  }
}
