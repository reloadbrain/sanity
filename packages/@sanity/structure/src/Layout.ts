export type Layout = 'default' | 'card' | 'media' | 'detail' | 'block'
export const LayoutOptions = ['default', 'card', 'media', 'detail', 'block']
