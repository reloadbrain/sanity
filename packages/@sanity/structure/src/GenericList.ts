import {StructureNode} from './StructureNodes'
import {ChildResolver} from './ChildResolver'
import {Layout, LayoutOptions} from './Layout'
import {MenuItem, MenuItemBuilder} from './MenuItem'
import {MenuItemGroup, MenuItemGroupBuilder} from './MenuItemGroup'
import {IntentChecker} from './Intent'

function maybeSerializeMenuItemGroup(item: MenuItemGroup | MenuItemGroupBuilder): MenuItemGroup {
  return item instanceof MenuItemGroupBuilder ? item.serialize() : item
}

function maybeSerializeMenuItem(item: MenuItem | MenuItemBuilder): MenuItem {
  return item instanceof MenuItemBuilder ? item.serialize() : item
}

function noChildResolver() {
  return undefined
}

export interface GenericList extends StructureNode {
  menuItems: MenuItem[]
  menuItemGroups: MenuItemGroup[]
  defaultLayout?: Layout
  canHandleIntent?: IntentChecker
  resolveChildForItem: ChildResolver
}

export type PartialGenericList = Partial<GenericList>

export abstract class GenericListBuilder<L extends PartialGenericList> {
  protected intentChecker?: IntentChecker

  constructor(protected spec: L) {}

  id(id: string) {
    this.spec.id = id
    return this
  }

  title(title: string) {
    this.spec.title = title
    return this
  }

  layout(layout: Layout) {
    if (!LayoutOptions.includes(layout)) {
      throw new Error(
        `\`layout\` must be one of ${LayoutOptions.map(item => `"${item}"`).join(', ')}`
      )
    }

    this.spec.defaultLayout = layout
    return this
  }

  menuItems(items: (MenuItem | MenuItemBuilder)[]) {
    this.spec.menuItems = items.map(maybeSerializeMenuItem)
    return this
  }

  menuItemGroups(groups: (MenuItemGroup | MenuItemGroupBuilder)[]) {
    this.spec.menuItemGroups = groups.map(maybeSerializeMenuItemGroup)
    return this
  }

  childResolver(resolver: ChildResolver) {
    this.spec.resolveChildForItem = resolver
    return this
  }

  canHandleIntent(checker: IntentChecker) {
    this.intentChecker = checker
    return this
  }

  serialize(): GenericList {
    return {
      id: this.spec.id || '',
      title: this.spec.title,
      type: 'genericList',
      defaultLayout: this.spec.defaultLayout,
      resolveChildForItem: this.spec.resolveChildForItem || noChildResolver,
      canHandleIntent: this.intentChecker,
      menuItems: this.spec.menuItems || [],
      menuItemGroups: this.spec.menuItemGroups || []
    }
  }
}
