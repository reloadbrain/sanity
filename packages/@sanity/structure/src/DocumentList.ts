import {StructureNode, EditorNode} from './StructureNodes'
import {GenericListBuilder, PartialGenericList, GenericList} from './GenericList'
import {ChildResolver, ChildResolverOptions} from './ChildResolver'
import {SortItem} from './Sort'

const resolveChildForItem: ChildResolver = (
  itemId: string,
  parent: StructureNode,
  options: ChildResolverOptions
): StructureNode | EditorNode | Promise<StructureNode | EditorNode> | undefined => {
  const parentItem = parent as DocumentList
  return {
    id: 'editor',
    type: 'document',
    options: {
      id: itemId,
      // @todo this aint always right
      type: parentItem.options.params && parentItem.options.params.type
    }
  }
}

export interface PartialDocumentList extends PartialGenericList {
  options?: DocumentListOptions
}

export interface DocumentList extends GenericList {
  options: DocumentListOptions
  resolveChildForItem: ChildResolver
}

interface DocumentListOptions {
  filter: string
  params?: {[key: string]: any}
  defaultOrdering?: SortItem[]
}

export class DocumentListBuilder extends GenericListBuilder<PartialDocumentList> {
  constructor(spec: PartialDocumentList = {}) {
    super(spec)
  }

  filter(filter: string): DocumentListBuilder {
    this.spec.options = {...(this.spec.options || {}), filter}
    return this
  }

  params(params: {}): DocumentListBuilder {
    this.spec.options = {...(this.spec.options || {filter: ''}), params}
    return this
  }

  defaultOrdering(ordering: SortItem[]): DocumentListBuilder {
    if (!Array.isArray(ordering)) {
      throw new Error('`defaultOrdering` must be an array of order clauses')
    }

    this.spec.options = {...(this.spec.options || {filter: ''}), defaultOrdering: ordering}
    return this
  }

  serialize(): DocumentList {
    if (typeof this.spec.id !== 'string' || !this.spec.id) {
      throw new Error('`id` is required for document lists')
    }

    if (!this.spec.options || !this.spec.options.filter) {
      throw new Error('`filter` is required for document lists')
    }

    return {
      ...super.serialize(),
      type: 'documentList',
      resolveChildForItem: this.spec.resolveChildForItem || resolveChildForItem,
      options: this.spec.options
    }
  }
}
