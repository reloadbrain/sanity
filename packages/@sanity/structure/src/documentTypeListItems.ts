import {Schema, defaultSchema} from './parts/schema'
import {dataAspects} from './parts/DataAspects'
import {ListItem, ListItemBuilder, SchemaType} from './ListItem'
import {DocumentList, DocumentListBuilder} from './DocumentList'
import {MenuItemBuilder} from './MenuItem'
import {DEFAULT_ORDERING_OPTIONS} from './Sort'
import {getPlusIcon, getSortIcon, getListIcon} from './parts/icon'
import {EditorBuilder} from './Editor'

const PlusIcon = getPlusIcon()
const SortIcon = getSortIcon()
const ListIcon = getListIcon()

export const getDocumentTypeListItems = (schema: Schema = defaultSchema): ListItem[] => {
  const resolver = dataAspects(schema)
  const types = resolver.getInferredTypes()

  return types.map(name => {
    const title = resolver.getDisplayName(name)
    const schemaType = schema.get(name)
    return getDocumentTypeListItem(name, title, schemaType)
  })
}

function getDocumentTypeListItem(name: string, title: string, type: SchemaType): ListItem {
  return new ListItemBuilder()
    .id(name)
    .title(title)
    .schemaType(type)
    .child(getDocumentList(name, title, type))
    .serialize()
}

function getDocumentList(name: string, title: string, type: SchemaType): DocumentList {
  return new DocumentListBuilder()
    .id(name)
    .title(title)
    .filter('_type == $type')
    .params({type: name})
    .defaultOrdering([]) // @todo
    .canHandleIntent(
      (intentName, params): boolean =>
        Boolean(intentName === 'edit' && params && params.id) ||
        Boolean(intentName === 'create' && params && params.type === name)
    )
    .menuItemGroups([
      {id: 'sorting', title: 'Sort'},
      {id: 'layout', title: 'Layout'},
      {id: 'actions', title: 'Actions'}
    ])
    .childResolver((documentId: string) =>
      new EditorBuilder()
        .id('editor')
        .type(name)
        .documentId(documentId)
        .serialize()
    )
    .menuItems([
      // Create new (from action button)
      new MenuItemBuilder()
        .title(`Create new ${title}`)
        .icon(PlusIcon)
        .intent({type: 'create', params: {type: name}})
        .showAsAction(true),

      // Sort by <Y>
      ...getOrderingMenuItemsForSchemaType(type),

      // Display as <Z>
      new MenuItemBuilder()
        .group('layout')
        .title('List')
        .icon(ListIcon)
        .action('setLayout')
        .params({layout: 'default'}),

      new MenuItemBuilder()
        .group('layout')
        .title('Details')
        .icon(ListIcon)
        .action('setLayout')
        .params({layout: 'detail'}),

      // Create new (from menu)
      new MenuItemBuilder()
        .group('actions')
        .title('Create newâ€¦')
        .icon(PlusIcon)
        .intent({type: 'create', params: {type: name}})
    ])
    .serialize()
}

function getOrderingMenuItemsForSchemaType(type: SchemaType) {
  return (type.orderings
    ? type.orderings.concat(DEFAULT_ORDERING_OPTIONS)
    : DEFAULT_ORDERING_OPTIONS
  ).map(ordering =>
    new MenuItemBuilder()
      .group('sorting')
      .title(`Sort by ${ordering.title}`)
      .icon(SortIcon)
      .action('setSortOrder')
      .params({by: ordering.by})
  )
}
