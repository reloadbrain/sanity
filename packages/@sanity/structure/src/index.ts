import {ListBuilder, ListInput} from './List'
import {getDocumentTypeListItems} from './documentTypeListItems'
import {MenuItemBuilder, MenuItem} from './MenuItem'
import {ListItemBuilder, ListItem} from './ListItem'
import {MenuItemGroup, MenuItemGroupBuilder} from './MenuItemGroup'
import {DocumentListBuilder, DocumentListInput} from './DocumentList'
import {EditorBuilder} from './Editor'
import {EditorNode} from './StructureNodes'

const StructureBuilder = {
  list: (spec?: ListInput): ListBuilder => new ListBuilder(spec),
  listItem: (spec?: ListItem): ListItemBuilder => new ListItemBuilder(spec),
  editor: (spec?: EditorNode): EditorBuilder => new EditorBuilder(spec),
  menuItem: (spec?: MenuItem): MenuItemBuilder => new MenuItemBuilder(spec),
  menuItemGroup: (spec?: MenuItemGroup): MenuItemGroupBuilder => new MenuItemGroupBuilder(spec),
  documentList: (spec?: DocumentListInput): DocumentListBuilder => new DocumentListBuilder(spec),
  documentTypeListItems: getDocumentTypeListItems
}

export {StructureBuilder}
