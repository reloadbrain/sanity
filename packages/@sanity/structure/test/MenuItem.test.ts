import {StructureBuilder as S} from '../src'

test('builds menu items with only title and action', () => {
  expect(S.menuItem({title: 'Foo', action: 'foo'}).serialize()).toMatchSnapshot()
})

test('throws if no title is set', () => {
  expect(() => S.menuItem({action: 'foo'}).serialize()).toThrowErrorMatchingSnapshot()
})

test('throws if neither action nor intent is set', () => {
  expect(() => S.menuItem({title: 'foo'}).serialize()).toThrowErrorMatchingSnapshot()
})

test('throws if setting both action AND intent', () => {
  expect(() =>
    S.menuItem()
      .title('foo')
      .action('foo')
      .intent({type: 'create'})
      .serialize()
  ).toThrowErrorMatchingSnapshot()
})

test('builds menu items with setters', () => {
  expect(
    S.menuItem()
      .title('Print')
      .action('print')
      .group('old-school')
      .icon(() => null)
      .params({foo: 'bar'})
      .showAsAction(false)
      .serialize()
  ).toMatchSnapshot()
})

test('builds menu item groups through constructor', () => {
  const input = {id: 'foo', title: 'Foo'}
  expect(S.menuItemGroup(input).serialize()).toMatchObject(input)
})

test('builds menu item groups through builder', () => {
  expect(
    S.menuItemGroup()
      .title('Foo')
      .id('foo')
      .serialize()
  ).toMatchObject({id: 'foo', title: 'Foo'})
})

test('throws if building menu item group without id', () => {
  expect(() =>
    S.menuItemGroup()
      .title('Foo')
      .serialize()
  ).toThrowError(/`id` is required/)
})

test('throws if building menu item group without title', () => {
  expect(() =>
    S.menuItemGroup()
      .id('foo')
      .serialize()
  ).toThrowError(/`title` is required/)
})
