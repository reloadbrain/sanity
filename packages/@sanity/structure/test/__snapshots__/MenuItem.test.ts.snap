// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds menu items with only title and action 1`] = `
Object {
  "action": "foo",
  "title": "Foo",
}
`;

exports[`builds menu items with setters 1`] = `
Object {
  "action": "print",
  "group": "old-school",
  "icon": [Function],
  "params": Object {
    "foo": "bar",
  },
  "showAsAction": false,
  "title": "Print",
}
`;

exports[`throws if neither action nor intent is set 1`] = `"\`action\` or \`intent\` required for menu item with title foo"`;

exports[`throws if no title is set 1`] = `"\`title\` is required for menu item"`;

exports[`throws if setting both action AND intent 1`] = `"cannot set both \`action\` AND \`intent\`"`;
